cranelift/fuzzgen/src/bin/check_opcode_signatures.rs

```diff
@@ -0,0 +1,140 @@
+    let mut insts = vec![];
+
+    for op in OPCODE_LIST {
+        if SKIPPED_OPCODES.contains(op) {
+            continue;
+        }
+
+        let constraints = op.constraints();
+
+        let ctrl_types = if let Some(ctrls) = constraints.ctrl_typeset() {
+            Vec::from_iter(TYPES.iter().copied().filter(|ty| ctrls.contains(*ty)))
+        } else {
+            vec![INVALID]
+        };
+
+        for ctrl_type in ctrl_types {
+            let rets = Vec::from_iter(
+                (0..constraints.num_fixed_results()).map(|i| constraints.result_type(i, ctrl_type)),
+            );
+
+            let mut cols = vec![];
+
+            for i in 0..constraints.num_fixed_value_arguments() {
+                match constraints.value_argument_constraint(i, ctrl_type) {
+                    ResolvedConstraint::Bound(ty) => cols.push(Vec::from([ty])),
+                    ResolvedConstraint::Free(tys) => cols.push(Vec::from_iter(
+                        TYPES.iter().copied().filter(|ty| tys.contains(*ty)),
+                    )),
+                }
+            }
+
+            let mut argss = vec![vec![]];
+
+            let mut cols = cols.as_slice();
+            while let Some((col, rest)) = cols.split_last() {
+                cols = rest;
+
+                let mut next = vec![];
+                for current in argss.iter() {
+                    for ty in col {
+                        let mut args = vec![*ty];
+                        args.extend_from_slice(&current);
+                        next.push(args);
+                    }
+                }
+
+                let _ = std::mem::replace(&mut argss, next);
+            }
+
+            for args in argss {
+                insts.push((*op, args, rets.clone()));
+            }
+        }
+    }
+
+    insts
+}
+
+#[derive(Eq, PartialEq, Debug)]
+struct Inst<'a> {
+    args: &'a [Type],
+    rets: &'a [Type],
+}
+
+fn build_sig_map<'a, T>(sigs: &'a [(Opcode, T, T)]) -> HashMap<Opcode, Vec<Inst<'a>>>
+where
+    T: AsRef<[Type]>,
+{
+    let mut insts = HashMap::<Opcode, Vec<Inst<'a>>>::default();
+
+    for (op, args, rets) in sigs {
+        insts.entry(*op).or_default().push(Inst {
+            args: args.as_ref(),
+            rets: rets.as_ref(),
+        });
+    }
+
+    insts
+}
+
+fn main() {
+    let all_ops = instruction_instantiations();
+    let everything = build_sig_map(&all_ops);
+    let fuzzed = build_sig_map(OPCODE_SIGNATURES);
+
+    let mut unknown = vec![];
+    for (op, insts) in fuzzed.iter() {
+        if let Some(known_insts) = everything.get(op) {
+            let invalid = Vec::from_iter(insts.iter().filter(|inst| !known_insts.contains(inst)));
+            if !invalid.is_empty() {
+                println!("# Invalid instantiations for Opcode::{:?}", op);
+                for inst in invalid {
+                    println!("- args: `{:?}`, rets: `{:?}`", inst.args, inst.rets);
+                }
+            }
```

Comment:
One question I have here is if this is the best way to report this information. Perhaps reworking this program as a test instead would be better, as that would prevent any backsliding on the existing signatures?

Reply:
Yeah, changing this to be a test would be awesome, that way we definitely wouldn't miss wrong signatures. (which there are way more than I would have thought 😅)

Reply:
I reworked it as a test, so you'll see the output if a new signature is added that doesn't match the constraints of the instruction DSL :+1: